### ISSUE TITLE
Semantics of dart team locks not well-defined

### ISSUE THREAD

**Issue by rkowalewski** [dash-project/dash#339]
[out.txt](https://github.com/dash-project/dash/files/865596/out.txt)
[env.txt](https://github.com/dash-project/dash/files/865600/env.txt)

The current implementation of DART locks to enable mutual exclusion for units in a team is implemented based on two MPI Windows:
Let us call them `Win_1` which stores the current holder of the lock and `Win_2` where units enqueue themselves in a linked list. The order in which ranks enqueue themselves determines the order in which they acquire the lock.

The operations to acquire a lock is as follows (adapted from the well known MCS lock):
~~~cpp
//1. atomically replace the current holder with myrank and fetch at the same time the rank of the current lock holder
int predecessor = MPI_Fetch_and_op(me, REPLACE, tail_pointer, Win_1);
flush(tail_unit, Win_1);
if (predecessor != UNIT_NULL) {
  //2. atomically add me as the next pointer in predecessor
  MPI_Fetch_and_op(me, REPLACE, next_pointer, Win_2);
  flush(predecessor, Win_2);
 //3. Block until predecessor passes me the lock
  MPI_Recv(predecessor);
}

/* It may be the case that the flush on Win_2 remotely completes before 
** the flush on Win_1. This is semantically valid in MPI-3 since there is no 
** happens before order across windows. It is some kind of magic how this 
** works but....it works.
*/
~~~

It is strictly required that the order of the operations is preserved as described above
Flushing a MPI window guarantees remote completion only with respect to operations on the specified window. If we execute two sequences of communication operations followed by a flush on two different MPI windows (as illustrated above), this does NOT (!!) guarantee that the communications on the first windows arrives before the operations on the second window. There is no happens before order across different windows.

Thus it may be the case that during the last operation (3), two processes deadlock because they wait for a matching `MPI_Send` from themselves.

I can reproduce this scenario on projekt03 with no thread support enabled and starting only with 2 MPI processes. See the attached files for environment and output on the top of this post.

So we should ensure that only the lock is implemented on a single MPI window. One processes allocates two elements (tail pointer and next pointer) while the other processes allocate 1 element (next pointer).
This is also more efficient since it decreases the memory footprint and window management overhead in MPI.

The issue is relevant both to #331 and #330 and thus should be included in Release 0.3.0. Or we completely revert the mutex stuff and postpone it to the next release. IMHO we need more unit tests for this feature in DART.

**Comment by devreal** [comment]
Mhh, but isn't the order given by the order of the flushes, i.e., 1) is guanranteed to  have finished before the `MPI_Fetch_and_op`  in 2) is called. 

>  IMHO we need more unit tests for this feature in DART.

Patches are welcome :)

**Comment by fuchsto** [comment]
Apart from that: hard tabs? Yakkkk.
https://github.com/dash-project/dash/blob/development/dart-impl/mpi/src/dart_synchronization.c

**Comment by devreal** [comment]
@fuchsto You hadn't seen that file before I touched it, did you? I already tried to get it in shape, will fix the rest...

**Comment by fuchsto** [comment]
I didn't blame you in a single syllable.
It rather looks like an emacs crime.

**Comment by fuchsto** [comment]
The plot thickens. No unit tests and emacs hard tabs. Elementary, Watson, the deed of a post-doc.

... seriously, though: the implementation of locks in dart-mpi was already broken before @devreal's changes and this has only been revealed by @fmoessbauer's introduction of `dash::Mutex`.
So if it cannot be fixed with reasonable effort (half a day max), I suggest to add it to "known limitations" and postpone the fix to 0.4.0.
I mean, it already is a limitation in 0.2.0, now it's a **known** limitation. Still an improvement.

**Comment by fuerlinger** [comment]
Agreed that it is probably best to postpone fixing this after 0.3.0!

**Comment by fuchsto** [comment]
Ha! The criminal returns to the scene of the crime!

The mutex type is currently only used by @fmoessbauer for his coarray implementation.
So if he is fine with this, so should we.

**Comment by rkowalewski** [comment]
> Mhh, but isn't the order given by the order of the flushes, i.e., 1) is guanranteed to  have finished before the `MPI_Fetch_and_op`  in 2) is called. 

No for exactly the reasons as discussed above.

>Agreed that it is probably best to postpone fixing this after 0.3.0!

OK I agree as well. Then we should revert the merge of DASH Mutex in #329.

**Comment by rkowalewski** [comment]
> The mutex type is currently only used by @fmoessbauer for his coarray implementation.
So if he is fine with this, so should we.

I do not think that @fmoessbauer is satisfied with a non-working solution. :)

**Comment by fuerlinger** [comment]
I think it's Huan's code - I would only consider accepting responsibility if the code also includes copious amounts of trailing whitespace.

**Comment by devreal** [comment]
>> Mhh, but isn't the order given by the order of the flushes, i.e., 1) is guanranteed to have finished before the MPI_Fetch_and_op in 2) is called.
>
> No for exactly the reasons as discussed above.

Now that is a fruitful discussion. 

> I think it's Huan's code - I would only consider accepting responsibility if the code also includes copious amounts of trailing whitespace.

I can put some in and re-add the tabs, if that helps to convince you :) (need to modify my editor though...)

**Comment by rkowalewski** [comment]
> Now that is a fruitful discussion. 

OK, I'll try to summarize. Yes, a flush guarantees remote completion between two subsequent communication operations. However, only within the same window and to the same target.

The code above (or the current implementation of locking) does something different. Suppose `unit 0 holds the current lock holder (tail pointer).
1. Apply an atomic operation to `unit 0` in `Win_1` in order to modify the tail pointer .
2. Apply an atomic operation to `predecessor` in `Win_2`.
3. Block until predecessor passes me the lock.

According to my understanding it is valid that operation 2 remotely completes before operation 1 because we have two different windows and two different targets. That is why there are no ordering guarantees.

Suggestions for improvements from my side are:
- Allocate everything only on one window.
- Probably replace the blocking receive with a a spin lock approach on the local memory since it should be faster than message passing. But this is not related to this issue.

I uploaded a [prototypical implementation](https://gist.github.com/rkowalewski/55203ddd730291d2d7ff6f07307cd653) of myself which I have tested some time ago.

**Comment by devreal** [comment]
Thanks, much appreciated. 

>  According to my understanding it is valid that operation 2 remotely completes before operation 1 because we have two different windows and two different targets. That is why there are no ordering guarantees.

That is where I am unsure. After each atomic operation we issue a flush, which is supposed to block until the operation is finished on the remote side. Hence, the order of operation 1 and 2, right? We cannot even issue operation 2 on the same process before operation 1 is remotely completed because the flush does not return earlier. Maybe my understanding of *block until remote completion* does not match the semantics of *block until remote completion* in the MPI standard?

I must admit I haven't tried to reproduce it, will take some time this afternoon. 

> Allocate everything only on one window.

Fine with me, although the overhead of having two windows should be small in the current implementation since the windows are not actually created but segments are attached.

**Comment by rkowalewski** [comment]
> That is where I am unsure. After each atomic operation we issue a flush, which is supposed to block until the operation is finished on the remote side. Hence, the order of operation 1 and 2, right? We cannot even issue operation 2 on the same process before operation 1 is remotely completed because the flush does not return earlier. Maybe my understanding of block until remote completion does not match the semantics of block until remote completion in the MPI standard?

The standard is vague what remote completion really means and MPI implementations have lots of freedom. One possible scenario is that MPI libraries have an intermediate buffer on the remote side to receive one-sided messages before they are finally transmitted to the target window. While the messages should arrive to the intermediate buffer in-order it may be the case that these intermediate buffers are flushed in reverse order. It is just an assumption and I do not know if this is really the case how it works. But it is possible in the MPI-3 Standard.

**Comment by fmoessbauer** [comment]
As I was mentioned multiple times in this ticket referring the Coarray: The `dash::Mutex` is not mandatory (or even necessary) for the `dash::Coarray<T>`, but for a `dash::Comutex`. As the Coarray stuff itself will definitely not be part of 0.3.0 I'm fine with pushing this to 0.4.0.

Another option would be to leave the DASH code as is, and mark it as experimental (or note the problems in DART).

If we want to revert the mutex PR, we also have to revert the documentation on codedocs.

**Comment by devreal** [comment]
The original problem has been addressed in PR #331. I think the explanation of the fault was not correct. Since we are using `MPI_Win_flush` the order in which the *atomic* operations on a process are performed is well defined. Of course, they may be interlocked in two processes but since they are atomic it does not matter. 

The overhead of using two different segments (not allocated windows) is marginal and thus can be ignored for now. We may want to look at that again once we overhauled the allocations in DART.
